protocols/autonatv2/src/client/handler.rs

```diff
@@ -0,0 +1,395 @@
+use std::{
+    cmp::min,
+    collections::VecDeque,
+    convert::identity,
+    i8::MAX,
+    io,
+    iter::{once, repeat},
+    sync::Arc,
+    task::Poll,
+    time::Duration,
+};
+
+use futures::{AsyncRead, AsyncWrite, AsyncWriteExt};
+use libp2p_core::{upgrade::ReadyUpgrade, Multiaddr};
+use libp2p_identity::PeerId;
+use libp2p_swarm::{
+    handler::{ConnectionEvent, FullyNegotiatedInbound, FullyNegotiatedOutbound, ProtocolsChange},
+    ConnectionHandler, ConnectionHandlerEvent, StreamProtocol, SubstreamProtocol,
+};
+use void::Void;
+
+use crate::{
+    client::ToBehaviour,
+    generated::structs::mod_DialResponse::ResponseStatus,
+    request_response::{
+        DialBack, DialDataRequest, DialDataResponse, DialRequest, DialResponse, Request, Response,
+    },
+    DIAL_BACK_PROTOCOL_NAME, REQUEST_PROTOCOL_NAME, REQUEST_UPGRADE,
+};
+use crate::{DATA_FIELD_LEN_UPPER_BOUND, DATA_LEN_LOWER_BOUND, DATA_LEN_UPPER_BOUND};
+
+use super::ResponseInfo;
+
+const DEFAULT_TIMEOUT: Duration = Duration::from_secs(10);
+const MAX_CONCURRENT_REQUESTS: usize = 10;
+
+#[derive(Debug, thiserror::Error)]
+pub(super) enum Error {
+    #[error("io error")]
+    Io(#[from] io::Error),
+    #[error("invalid data request index: {index} (max: {max})")]
+    InvalidDataRequestIndex { index: usize, max: usize },
+    #[error("data request too large: {len} (max: {max})")]
+    DataRequestTooLarge { len: usize, max: usize },
+    #[error("data request too small: {len} (min: {min})")]
+    DataRequestTooSmall { len: usize, min: usize },
+    #[error("timeout")]
+    Timeout(#[from] futures_bounded::Timeout),
+    #[error("invalid nonce: {nonce} (provided by (by))")]
+    WrongNonceGiven { nonce: u64, by: PeerId },
+    #[error("request protocol removed by peer")]
+    RequestProtocolRemoved,
+    #[error("dial back protocol removed by me")]
+    DialBackProtocolRemoved,
+}
+
+pub(super) struct Handler {
```

Comment:
Perhaps we should make two different `Handler`s:

- `Handler`s are unique per connection, meaning the handler that receives the `dial-back` message **is not** the same `Handler` instance as the one that sent the original `dial-request` message.
- If we make one `Handler` that only supports the inbound `/dial-back` protocol and one `Handler` that only perform the outbound `/dial-request` protocol, this distinction is much clearer.
- We can compose these `Handler`s in our `Behaviour` using `Either`.
	- The `dial_back::Handler` is only relevant for inbound connections and never opens a stream (i.e. `OutboundProtocol = DeniedUpgrade;`)
	- The `dial_request::Handler` is only relevant for outbound connections and never receives any streams (i.e. `InboundProtocol = DeniedUpgrade;`).

This should allow us to separate the state more clearly. What do you think?

I think the `dial_back::Handler` for example would be incredible simple. All it does is answer to incoming streams, read the nonce and report it back to the behaviour. The behaviour should have the state about which `/dial-request`s it initiated and can match it back with a received response from the other handler.

Similarly, the `dial_request::Handler` needs only do the following things:

- Check for the `/dial-request` protocol for each connection and report the availability of an autonat server. The behaviour can then store this connection as an available autonat server connection.
- Receive instructions to make a `/dial-request` on a particular connection, open the stream, do the request-response dance and report the result back to the behaviour. We can use the nonce as the "ID" for the `FuturesMap` and can thus always identify the outcome of the `/dial-request`. Once completed, the handler reports the result back to the behaviour.

The behaviour then needs to aggregate the state from both handlers.

Reply:
That's addressed.